### PR TITLE
Globalization of gls updates

### PR DESCRIPTION
**Pull request by AitorBengoechea**
The aim of this branch is to make the GLS that sandy has fully global (also, for constrained gls update). To do this, the `gls_update` methods have been updated to work with sensitivity vectors. @EnricaBelfiore will implement the changes and test the functions with sensitivity vectors.

### CODE REVIEW

**Comment by AitorBengoechea** [comment]
Future work: Once the implementation is done, test the fy.py methods and update if it is necessary

**Comment by EnricaBelfiore** [comment]
Some comments on the last commit:

- With the new implementation of the GLS update, the design matrix S can be MXN with M that could be also equals to 1. I specified it in the docstring but maybe it is redundant. What do you think?
- The `sandy.constrained_gls_update()` function and all the private functions to perform it are removed. This is because, in terms of implementation, the GLS update formula and constrained GLS update formula are equivalent if Vy_extra is zero (default is `None`). 
- I inserted a warning in the `sandy.gls._gls_G_inv()` function. This is because when the determinant of the matrix `S*V_prior*S^T + Vy_extra` is close to zero (e.x. 1e-19) the method `np.linalg.inv()` will not work properly. Serching on internet I found that in this case it is possible to perform the pseudo-inverse of a matrix with the method `np.linalg.pinv()` ([[inverse matrix np.linalg.inv in Python, gives weird results, why?](https://stackoverflow.com/questions/66094987/inverse-matrix-np-linalg-inv-in-python-gives-weird-results-why)](https://stackoverflow.com/questions/66094987/inverse-matrix-np-linalg-inv-in-python-gives-weird-results-why)). Is it more convinient to use the latter in that case?
- The tests in the `sandy.fy.py` script are wrong. I kept them as they are but they will be fixed in another pull request.

**Review by luca-fiorito-11 — CHANGES_REQUESTED** [review]
Just minor changes

File: sandy/fy.py [hunk 1]
```diff
@@ -895,7 +895,11 @@ def ishikawa_factor(self, zam, e, Vy_extra,
             model_sensitivity_object = decay_data
         S = _gls_setup(model_sensitivity_object, kind)
         # Perform Ishikawa factor:
-        ishikawa = sandy.ishikawa_factor(S, Vx_prior, Vy_extra)
+        Vy_extra_ = pd.DataFrame(Vy_extra)
+        index = Vy_extra_.index
+        S = S.reindex(index=index, columns=Vx_prior.index).fillna(0)
+        ishikawa = sandy.ishikawa_factor(S.values, Vx_prior.values, Vy_extra_.values)
```
[the remark this answers is not in the document]
Comment (EnricaBelfiore) [reply sandy/fy.py:901]: This will be done in another PR

File: sandy/gls.py [hunk 2]
```diff
@@ -22,7 +21,7 @@
         ]
 
 x_prior = [1, 2, 3]
-y_extra = pd.Series([2, 3, 4], index=[1, 2, 3])
+y_extra = np.array([2, 3, 4])
```
Comment (luca-fiorito-11) [inline sandy/gls.py:24]: @EnricaBelfiore, what are all these things? If not needed we can remove them

File: sandy/core/cov.py [hunk 3]
```diff
@@ -1962,6 +1958,73 @@ def sparse_tables_inv(a, rows=1000):
     return invert_matrix
 
 
+def segmented_pivot_table(data_stack, index, columns, values, rows=10000000):
```
Comment (luca-fiorito-11) [inline sandy/core/cov.py:1961]: @EnricaBelfiore , do we use this?

File: sandy/core/cov.py [hunk 4]
```diff
@@ -1885,7 +1881,7 @@ def sparse_tables_dot(a, b, rows=1000):
     return dot_product
 
 
-def sparse_tables_dot_multiple(matrix_list, rows=1000):
+#def sparse_tables_dot_multiple(matrix_list, rows=1000):
```
Comment (luca-fiorito-11) [inline sandy/core/cov.py:1884]: @EnricaBelfiore , why commenting it and not removing it?

**Review by luca-fiorito-11 — APPROVED** [review]

File: sandy/gls.py (hunk 2 above)
Comment (EnricaBelfiore) [reply sandy/gls.py:24]: They have been inserted and used in the docstring example of `sandy.gls._y_calc()`. They will be removed (and the function as well) in another PR

File: sandy/core/cov.py (hunk 3 above)
Comment (EnricaBelfiore) [reply sandy/core/cov.py:1961]: Yes, it is used in the class method `sandy.CategoryCov.from_stack()`

File: sandy/core/cov.py (hunk 4 above)
Comment (EnricaBelfiore) [reply sandy/core/cov.py:1884]: My fault, I forgot to delete the comments but the function is no more used and it will be removed

**Review by luca-fiorito-11 — APPROVED** [review]

**Pull request merged**